air/src/options.rs

```diff
@@ -34,6 +36,157 @@ pub enum FieldExtension {
     Cubic = 3,
 }
 
+/// Represents the FRI folding factor option, a power of two
+#[repr(u8)]
+pub enum FriFoldingFactor {
```

Comment:
I am not really a fan of this approach, which I feel is too heavy for what it aims at achieving. For instance supporting a different folding factor (eg 2), would require multiple changes across the library, plus break the existing ordering (as First would become 2 here, and shift all the other values).

Reply:
@Nashtare As it currently stands, `2` isn't a legal value, and the point of the enum is to ensure that you couldn't even try an illegal value if you wanted to do so.  From that standpoint, wanting to make `2` legal seems like it _should_ be a breaking change.

Another way of looking at it is that, if we wanted it to be _easy_ to assume a currently illegal value in the near future, then this proposal invites the question of the wisdom of _panicking_ on such a value in the first place, as opposed to something lighter, like a warning.

Reply:
> As it currently stands, 2 isn't a legal value, and the point of the enum is to ensure that you couldn't even try an illegal value if you wanted to do so. From that standpoint, wanting to make 2 legal seems like it should be a breaking change.

I don't agree on this. If we were to support additional folding factors in the future, in no way this would be a breaking change. In the sense that code successfully compiling with winterfell as backend would keep on behaving correctly. Handling a previously erroneous case such that it is in the future a valid possibility is not a breaking change, but an extension of existing feature.

In addition (see my main review comment), we can still handle the panics by replacing the asserts with equivalent conditional early error returns (similarly to what you've done on `ProofOptions::new()`, without needing all those enums.

Reply:
@Nashtare Truth be told, I wish I had not needed to resort to the conditional error returns for any parameter.  The problem with the grinding factor and the query number is that there are still too many legal values to restrain those to an enum.  Ideally, the fewer legal states can be accepted, the less states there are to consider for coverage.

Can I ask what your issue with enums is here?  For me, enums increase readability, and so I tend to prefer more to less, but it seems to be the reverse for you?